### DISCUSSION TITLE
Netty pooled allocator soft limits

### DISCUSSION THREAD

**Discussion by franz1981** [netty/netty#11845] [Ideas]
Similarly to https://openjdk.java.net/jeps/346 / https://openjdk.java.net/jeps/8204088 / `-XX:SoftMaxHeapSize` (see https://bugs.openjdk.org/browse/JDK-8222145 too, which is unrelated to the soft limit, but a required detail for this Netty feature, instead) , it would be nice if Netty pooled allocator would be able to return back some memory to the OS based on some soft limit and time/load/capacity configurable policies.
This would be very useful for containerized applications (or mild loaded ones): jemalloc as well is able to uncommit memory back to the OS under certain conditions, but I didn't dig further how much it looks like the feature I'm proposing.

Comments/ideas are wellcome!

NOTES:
right now I see that only huge allocations and finalized arenas benefit from being released/releasable.
An easy approach, although less precise, to solve this would be that when total allocations exceed a soft limit, the added chunk should be created as "unpooled" (meaning that would be released and uncommitted when not used/referenced anymore); in addition a global tick + lease renew on usage can help deciding when is time to release it for real, in case a max-idle-time is configured for the memory allocated beyond the soft limit. While half-dead (eligible for GC), the chunk is still available to be (re)used to cope with incoming load.

**Comment by franz1981** [comment]
The real struggle here is that even calling `free` won't return anything back till `trimming` on OS allocator won't happen: which means that there's some external help required here to make it to work

**Reply by tstuefe** [reply]
the new auto trim feature in jdk17+ could help here

**Reply by franz1981** [reply]
Adding https://bugs.openjdk.org/browse/JDK-8293114 for reference

@normanmaurer this can be helpful, wdyt?

**Reply by franz1981** [reply]
https://github.com/netty/netty/pull/8941 is currently allowing trimming in a timely way, but will basically cause unused arenas from some time to be recycled (at https://github.com/netty/netty/blob/7dba4fcf499425914e09993d1f2bd439794cb3d6/buffer/src/main/java/io/netty/buffer/PoolThreadCache.java#L441), which won't still make them to hit https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PoolArena.java#L721 and actually release anything back to the OS, am I correct @chrisvest / @normanmaurer ?
